ols/utils/auth_dependency.py

```diff
@@ -0,0 +1,210 @@
+"""Manage authentication flow for FastAPI endpoints with K8S/OCP."""
+
+import logging
+
+import kubernetes.client
+from fastapi import HTTPException, Request
+from kubernetes.client.rest import ApiException
+
+from ols.utils import config
+
+logger = logging.getLogger(__name__)
+
+
+class K8sClientSingleton:
+    """Return the Kubernetes client instances.
+
+    Ensures we initialize the k8s client only once per application life cycle.
+    manage the initialization and config loading.
+    """
+
+    _instance = None
+    _api_client = None
+    _auth_api = None
+    _authorization_api = None
+
+    def __new__(cls):
+        """Create a new instance of the singleton, or returns the existing instance.
+
+        This method initializes the Kubernetes API clients the first time it is called.
+        and ensures that subsequent calls return the same instance.
+        """
+        if cls._instance is None:
+            cls._instance = super(K8sClientSingleton, cls).__new__(cls)
+            configuration = kubernetes.client.Configuration()
+
+            try:
+                try:
+                    kubernetes.config.load_kube_config(
+                        client_configuration=configuration
+                    )
+                except Exception:
+                    try:
+                        logger.warn(
+                            "kubeconfig file not found, trying in-cluster configuration"
+                        )
+                        kubernetes.config.load_incluster_config(
+                            client_configuration=configuration
+                        )
+                    except Exception:
+                        logger.warn(
+                            "Failed to load in-cluster Kubernetes configuration"
+                        )
+                        logger.warn("Trying to load k8s config from app config")
+                        if (
+                            config.ols_config.authentication_config.k8s_cluster_api
+                            and config.dev_config.k8s_auth_token
+                        ):
+                            configuration.api_key["authorization"] = (
+                                config.dev_config.k8s_auth_token
+                            )
+                            configuration.api_key_prefix["authorization"] = "Bearer"
```

Comment:
i think this order is backwards.  If the ols config supplied a k8s token, that is what we should prefer to use.  If that is not provided in the ols config, then we should attempt to load the in cluster config from the pod (which is what i'd expect to end up happening most of the time, the ols config would only be for override or dev/debug purposes)

Reply:
also it's not clear to me that regardless of what exception is thrown while loading the k8s config, that setting this field and proceeding is an appropriate action to to take.  If you think it is, please comment this code explaining why it's appropriate

Reply:
changed the order and added the exact expected exception